python/lsst/ap/association/utils.py

```diff
@@ -0,0 +1,72 @@
+# This file is part of ap_association.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Utilities for working with the APDB.
+"""
+__all__ = ("convertTableToSdmSchema",)
+
+import numpy as np
+import pandas as pd
+
+import lsst.dax.apdb as daxApdb
+
+
+def convertTableToSdmSchema(apdbSchema, sourceTable,
+                            tableId=daxApdb.ApdbTables.DiaSource):
+    """Force a table to conform to the schema defined by the APDB.
+
+    This method uses the table definitions in ``sdm_schemas`` to
+    load the schema of the APDB, and does not actually connect to the APDB.
+
+    Parameters
+    ----------
+    apdbSchema : `lsst.dax.apdb.apdbSchema.ApdbSchema`
+        Schema from ``sdm_schemas`` containing the table definition to use.
+    sourceTable : `pandas.DataFrame`
+        The input table to convert.
+    tableId : `sqlalchemy.schema.Table`, optional
+        Database table identifier for the given schema.
+        If not set, defaults to the DiaSource table.
+
+    Returns
+    -------
+    `pandas.DataFrame`
+        A table with the correct schema for the APDB and data copied from
+        the input ``sourceTable``.
+    """
+    table = apdbSchema.tableSchemas[tableId]
+
+    data = {}
+    nSrc = len(sourceTable)
+    for columnDef in table.columns:
+        if columnDef.name in sourceTable.columns:
+            dtype = apdbSchema.column_dtype(columnDef.datatype)
+            data[columnDef.name] = pd.Series(sourceTable[columnDef.name], dtype=dtype)
+        else:
+            dataInit = np.zeros(nSrc, dtype=apdbSchema.column_dtype(columnDef.datatype))
+            if columnDef.nullable:
+                try:
+                    dataInit.fill(None)
+                except (TypeError, ValueError):
+                    pass
```

Comment:
For efficiency it may be better to skip all-NULL columns completely. For SQL it will save some CPU and bandwidth by not sending default values. For Cassandra it may be even more important, as Cassandra stores NULLs as "tombstones" and too many tombstones can result in problems.

Reply:
That would actually be helpful for a change I want to make later. Would it be better for those columns to be missing when the tables are supplied to `apdb.store()` in `diaPipe`, or should we provide a list of columns to skip?

Reply:
I think it would be simpler just to omit them from datframes.

Reply:
Particularly if you are omitting missing (nullable) columns, I think you can do this without looping and creating a bunch of new `pd.Series`.   You just need a dict of `coldef_map = {'column_name': dtype}`

and then you can return `sourcetable.as_type(coldef_map)` which I think should be more efficient.

Reply:
A question which @andy-slac might be able to answer is whether we can instead get the dtypes from `felis.datamodel`, which would allow us to work directly from `sdm_schemas` and not (ab)use the APDB class for type mapping.   Presumably it would make it easier for us to handle the DRP transformations as well.

Reply:
I've pushed a change that duplicates a small amount of code from `dax_apdb`, which separates the SDM type checking from the APDB class.